text/0000-stack-buildpacks.md

```diff
@@ -0,0 +1,668 @@
+
+ ```bash
+#!/usr/bin/env bash
+
+apt update -y
+
+for package in $(cat $3 | yj -t | jq -r ".entries | .[] | .mixin"); do
+  apt install $package
+done
+
+cat << EOF > $1/stack-layer.toml
+[[excludes]]
+paths = [ "/var/cache" ]
+cache = true
+EOF
+```
+
+Based on the proposal in [App Mixins](https://github.com/buildpacks/rfcs/pull/112), the user's `project.toml` might include:
+
+```toml
+[build]
+mixins = [ "libpq" ]
+```
+
+This would instruct the Apt Buildpack to install the libpq package.
+
+## Example: CA Certificate Buildpack
+
+Support for custom CA Certificates can be accomplished with two buildpacks: a stackpack that can install the cert, and a normal buildpack that can provide a cert in the build plan.
+
+### CA Cert Installer Stackpack
+
+A buildpack that installs custom CA Certificates would have a `buildpack.toml` that looks like this:
+
+```toml
+[buildpack]
+id = "example/cacerts"
+privileged = true
+
+[[stacks]]
+id = "io.buildpacks.stacks.bionic"
+```
+
+Its `bin/detect` would have the following contents:
+
+ ```bash
+#!/usr/bin/env bash
+
+cat << EOF > $2
+[[provides]]
+name = "cacert"
+EOF
+
+exit 0
+```
+
+Its `bin/build` would would install each `cacert` in the build plan. It would have the following contents:
+
+ ```bash
+#!/usr/bin/env bash
+
+# filter to the cert
+for file in $(cat $3 | yj -t | jq -r ".entries | .[] | select(.name==\"cacert\") | .metadata | .file"); do
+  cert="$(cat $3 | yj -t | jq -r ".entries | .[] | select(.name==\"cacert\") | .metadata | select(.file==\"$file\") | .content")"
+  echo $cert > /usr/share/ca-certificates/$file
+  chmod 644 /usr/share/ca-certificates/$file
+done
+
+update-ca-certificates
+```
+
+### CA Cert Provider Buildpack
+
+The stackpack must be used with a buildpack that provides a certificate(s) for it to install. That buildpack would have the following `buildpack.toml`:
+
+```toml
+[buildpack]
+id = "my/db-cert"
+
+[[stacks]]
+id = "io.buildpacks.stacks.bionic"
+```
+
+Its `bin/detect` would require a certificate with the following contents:
+
+ ```bash
+#!/usr/bin/env bash
+
+cat << EOF > $2
+[[requires]]
+name = "cacert"
+
+[requires.metadata]
+file = "database.crt"
+content = """
+$(cat $CNB_BUILDPACK_DIR/database.crt)
+"""
+
+[[requires]]
+name = "cacert"
+
+[requires.metadata]
+file = "server.crt"
+content = """
+$(cat $CNB_BUILDPACK_DIR/server.crt)
+"""
+EOF
+
+exit 0
+```
+
+Its `bin/build` would do nothing, and `have the following contents:
+
+ ```bash
+#!/usr/bin/env bash
+
+exit 0
+```
+
+## Example: jq mixin
+
+A buildpack may that requires the `jq` package may have it provided by either a stack, stack buildpack, or application buildpack.
+
+```toml
+[[or.requires]]
+mixin = "jq"
+
+[[or.requires]]
+name = "jq"
+```
+
+## Future work
+
+In the future, we plan to enhance the stack buildpack interface with the following:
+
+* A `CNB_STACK_TYPE` env var that a stack buildpack can use to behave differently on each part of the stack
```

Comment:
```suggestion
```
I would just take this out because I am not convinced we will want or need this eventually, but it's in future work so I don't care too much.

Reply:
I'd like to leave it, because it's come up as a question a few times. So I'd like to at least call out that we _aren't_ doing it in this iteration (if at all)

Reply:
Addressed in 439b4ec